conn.go

```diff
@@ -186,33 +189,30 @@ func Connect(addr string, cfg *ConnConfig, errorHandler ConnErrorHandler, sessio
 	}
 
 	c := &Conn{
-		conn:         conn,
-		r:            bufio.NewReader(conn),
-		cfg:          cfg,
-		uniq:         make(chan int, streams),
-		calls:        make(map[int]*callReq),
-		timeout:      cfg.Timeout,
-		version:      uint8(cfg.ProtoVersion),
-		addr:         conn.RemoteAddr().String(),
-		errorHandler: errorHandler,
-		compressor:   cfg.Compressor,
-		auth:         cfg.Authenticator,
-		headerBuf:    make([]byte, headerSize),
-		quit:         make(chan struct{}),
-		session:      session,
-		numStreams:   streams,
-	}
+		conn:             conn,
+		r:                bufio.NewReader(conn),
+		cfg:              cfg,
+		availableStreams: int32(maxStreams - 1), // reserve stream 0
+		streamCounter:    0,
+		calls:            make(map[int]*callReq),
+		timeout:          cfg.Timeout,
+		version:          uint8(cfg.ProtoVersion),
+		addr:             conn.RemoteAddr().String(),
+		errorHandler:     errorHandler,
+		compressor:       cfg.Compressor,
+		auth:             cfg.Authenticator,
+		headerBuf:        make([]byte, headerSize),
+		quit:             make(chan struct{}),
+		session:          session,
+		numStreams:       streams,
+		maxStreams:       maxStreams,
+	}
+	c.streamWaiter = sync.NewCond(&c.mu)
```

Comment:
I dont think its a good idea to use the same mutex as the inflight map, as that would prevent incoming responses being dispatched while requests are searching for a stream